src/aspire/utils/bot_align.py

```diff
@@ -0,0 +1,181 @@
+"""
+This code is an ASPIRE port of Ruiyi Yang's `Bayesian Optimal Transport Alignment`.
+
+`https://github.com/RuiyiYang/BOTalign`
+
+`https://arxiv.org/pdf/2305.12310.pdf`
+"""
+
+import numpy as np
+import pymanopt
+from numpy.linalg import norm
+from scipy.optimize import minimize
+
+from aspire.operators import wemd_embed
+from aspire.utils.rotation import Rotation
+
+
+def align_BO(
+    vol_ref,
+    vol_given,
+    loss_type="wemd",
+    downsampling_level=32,
+    max_iters=200,
+    refine=True,
+    tau=1e-3,
+    man_max_iter=500,
+    man_min_grad=0.1,
+    man_min_sz=0.1,
+    verbosity=0,
+    dtype=None,
+):
+    """
+    This function returns a rotation matrix R that best aligns vol_ref with the rotated version of vol_given.
+
+    :param vol_ref: The reference `Volume`
+    :param vol_given: The `Volume` to be aligned
+    :param loss_type: 'wemd' or 'eu'. Default 'wemd'.
+        If the heterogeneity between vol_ref and vol_given is high, 'eu' is recommended.
+    :param downsampling_level: Downsampling (pixels). Integer, defaults 32.
+        If alignment fails try larger values.
+    :param max_iters: Maximum iterations. Integer, defaults 200.
+        If alignment fails try larger values.
+    :param refine: Whether to perform refinement. Boolean, defaults True.
+    :param tau: Regularization parameter for surrogate problems. Numeric, defaults 1e-3.
+    :param man_max_iter: Stopping criterion for surrogate problems--maximum iterations. Interger, defaults 500.
+    :param man_min_grad: Stopping criterion for surrogate problems--minimum gradient norm. Numeric, defaults 0.1.
+    :param man_min_sz: Stopping criterion for surrogate problems--minimum step size. Numeric, defaults 0.1.
+    :param verbosity: Whether to print surrogate problem optimization details. Boolean, defaults 0.
+    :param dtype: Numeric dtype to perform computations with.
+        Default `None` infers dtype from `vol_ref`.
+    :return: two rotation matrices R_init (without refinement) and R_est (with refinement).
+        If refine = False, then R_est is set equal to R_init.
+    """
+    # Infer dtype
+    dtype = np.dtype(dtype or vol_ref.dtype)
+
+    # Convert volume data dtype if needed.
```

Comment:
Do we want to warn/error for dtype mismatch here? This is the usual policy IIRC.

Reply:
sure